File: apps/backend-java/src/main/java/com/luban/backend/service/UserService.java
package com.luban.backend.service;

import com.luban.backend.dto.UserListResponse;
import com.luban.backend.dto.UserResponse;
import com.luban.backend.entity.User;
import com.luban.backend.exception.BusinessException;
import com.luban.backend.mapper.UserMapper;
import org.springframework.dao.DataIntegrityViolationException;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import java.time.Instant;
import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

@Service
public class UserService {

    private final UserMapper userMapper;
    private final PasswordEncoder passwordEncoder;

    public UserService(UserMapper userMapper, PasswordEncoder passwordEncoder) {
        this.userMapper = userMapper;
        this.passwordEncoder = passwordEncoder;
    }

    public UserListResponse list(int page, int size, String keyword) {
        if (page <= 0) page = 1;
        if (size <= 0) size = 10;
        int offset = (page - 1) * size;
        String kw = (keyword != null && !keyword.isBlank()) ? keyword.trim() : null;
        List<User> users = userMapper.list(kw, offset, size);
        long total = userMapper.count(kw);
        List<UserResponse> list = users.stream().map(UserResponse::fromEntity).collect(Collectors.toList());
        return new UserListResponse(list, total);
    }

    public UserResponse get(String id) {
        User u = userMapper.getById(id);
        if (u == null) throw BusinessException.userNotFound();
        return UserResponse.fromEntity(u);
    }

    public UserResponse create(String username, String password, String name, String role) {
        if (role == null || role.isBlank()) role = "user";
        if (password == null || password.isBlank()) throw BusinessException.invalidArgument("password required");
        String hash = passwordEncoder.encode(password);
        User u = new User();
        u.setId(UUID.randomUUID().toString());
        u.setUsername(username);
        u.setName(name != null ? name : "");
        u.setRole(role);
        u.setStatus("active");
        u.setPassword(hash);
        Instant now = Instant.now();
        u.setCreatedAt(now);
        u.setUpdatedAt(now);
        try {
            userMapper.insert(u);
        } catch (DataIntegrityViolationException e) {
            if (e.getMessage() != null && e.getMessage().contains("Duplicate")) {
                throw BusinessException.usernameConflict();
            }
            throw e;
        }
        return UserResponse.fromEntity(u);
    }

    public UserResponse update(String id, String username, String name, String role, String status, String password) {
        User u = userMapper.getById(id);
        if (u == null) throw BusinessException.userNotFound();
        if (username != null) u.setUsername(username);
        if (name != null) u.setName(name);
        if (role != null) u.setRole(role);
        if (status != null) u.setStatus(status);
        u.setUpdatedAt(Instant.now());
        try {
            int n = userMapper.update(u);
            if (n == 0) throw BusinessException.userNotFound();
        } catch (DataIntegrityViolationException e) {
            if (e.getMessage() != null && e.getMessage().contains("Duplicate")) {
                throw BusinessException.usernameConflict();
            }
            throw e;
        }
        if (password != null && !password.isBlank()) {
            userMapper.updatePassword(id, passwordEncoder.encode(password), u.getUpdatedAt());
        }
        return UserResponse.fromEntity(userMapper.getById(id));
    }

    public UserResponse updateStatus(String id, String status) {
        User u = userMapper.getById(id);
        if (u == null) throw BusinessException.userNotFound();
        Instant now = Instant.now();
        int n = userMapper.updateStatus(id, status, now);
        if (n == 0) throw BusinessException.userNotFound();
        return UserResponse.fromEntity(userMapper.getById(id));
    }
}

File: apps/backend-go/internal/service/user_service.go
package service

import (
	"context"

	"github.com/google/uuid"
	"golang.org/x/crypto/bcrypt"

	"github.com/xiaoshuai1024/luban-backend-go/internal/model"
	"github.com/xiaoshuai1024/luban-backend-go/internal/repository"
)

type UserService struct {
	repo *repository.UserRepository
}

func NewUserService(repo *repository.UserRepository) *UserService {
	return &UserService{repo: repo}
}

func (s *UserService) List(ctx context.Context, page, size int, keyword string) ([]*model.User, int, error) {
	return s.repo.List(ctx, page, size, keyword)
}

func (s *UserService) Get(ctx context.Context, id string) (*model.User, error) {
	return s.repo.Get(ctx, id)
}

func (s *UserService) Create(ctx context.Context, username, password, name, role string) (*model.User, error) {
	if role == "" {
		role = "user"
	}
	if password == "" {
		return nil, ErrInvalidCredentials
	}
	hash, err := bcrypt.GenerateFromPassword([]byte(password), bcrypt.DefaultCost)
	if err != nil {
		return nil, err
	}
	u := &model.User{
		ID:       uuid.NewString(),
		Username: username,
		Name:     name,
		Role:     role,
		Status:   "active",
		Password: string(hash),
	}
	if err := s.repo.Create(ctx, u); err != nil {
		return nil, err
	}
	return u, nil
}

// Update 更新用户基本信息，并可选地修改密码。
//
// 设计原因：旧实现把 password 列放进通用 Update SQL，而调用方组装 model.User
// 时通常不填 Password，导致每次 PUT 用户都把密码清空（数据损坏 Bug 2）。
// 现在 Update SQL 不再含 password 列；仅当 newPassword 非空时，单独走
// UpdatePassword 流程（先 bcrypt 再更新），与 Java UserMapper.update 行为一致。
//
// newPassword 为空字符串表示不改密（保持原密码不变）。
func (s *UserService) Update(ctx context.Context, u *model.User, newPassword string) error {
	if err := s.repo.Update(ctx, u); err != nil {
		return err
	}
	if newPassword != "" {
		hash, err := bcrypt.GenerateFromPassword([]byte(newPassword), bcrypt.DefaultCost)
		if err != nil {
			return err
		}
		if err := s.repo.UpdatePassword(ctx, u.ID, string(hash)); err != nil {
			return err
		}
		// 回填内存对象，便于上层返回一致状态（Password 字段 json:"-" 不外泄）。
		u.Password = string(hash)
	}
	return nil
}

